### ISSUE TITLE
PanacheEntityBase: support for Hibernate @Filter

### ISSUE THREAD

**Issue by fiorenzino** [quarkusio/quarkus#9036]
**Description**
In panache you can use hibernate @Filter(). But to activate filters, you must:
- call the entity manager and unwrap to hibernate session
- enableFilter by name
- setParameter with name and value

**Implementation ideas**
We could add to 2 methods to enable a filter with parameter/parameters:

```
public static void enableFilter(String filterName, String parameterName, Object value) {
        JpaOperations.getEntityManager().unwrap(Session.class)
                .enableFilter(filterName)
                .setParameter(parameterName, value);
    }

    public static void enableFilter(String filterName, Map<String, Object> params) {
        JpaOperations.getEntityManager().unwrap(Session.class)
                .enableFilter(filterName);
        for (Map.Entry<String, Object> entry : params.entrySet()) {
            JpaOperations.getEntityManager().unwrap(Session.class)
                    .getEnabledFilter(filterName)
                    .setParameter(entry.getKey(), entry.getValue());
        }
    }
```
And a new method to find() using the filters (implemented at build time):
 ```
public static PanacheQuery<MobilePhone> findByFilters(String orderBy) {
        if (orderBy != null) {
            return MobilePhone.find("select a from MobilePhone a", Sort.by(orderBy));
        }
        return MobilePhone.find("select a from MobilePhone a");
    }
```

My complete use case:
2 entities and 1 rest class, to activate different filters by query parameters:

```
import io.quarkus.hibernate.orm.panache.PanacheEntity;
import io.quarkus.hibernate.orm.panache.PanacheQuery;
import io.quarkus.hibernate.orm.panache.runtime.JpaOperations;
import io.quarkus.panache.common.Sort;
import org.hibernate.Session;
import org.hibernate.annotations.Filter;
import org.hibernate.annotations.FilterDef;
import org.hibernate.annotations.ParamDef;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Entity
@Table(name = "persons")
@FilterDef(name = "lastName", parameters = @ParamDef(name = "lastName", type = "string"))
@Filter(name = "lastName", condition = "lastName LIKE :lastName")

@FilterDef(name = "tax_code", parameters = @ParamDef(name = "tax_code", type = "string"))
@Filter(name = "tax_code", condition = "tax_code = :tax_code")

@FilterDef(name = "phones_by_brand", parameters = @ParamDef(name = "brand", type = "string"))
@FilterDef(name = "phones_by_phone_model", parameters = @ParamDef(name = "phone_model", type = "string"))
@FilterDef(name = "phones_by_mobile_operator", parameters = @ParamDef(name = "mobile_operator", type = "string"))

public class Person extends PanacheEntity {

    public String firstName;
    public String lastName;
    public String tax_code;

    @OneToMany(mappedBy = "owner", fetch = FetchType.EAGER, cascade = CascadeType.ALL)
    @Filter(name = "phones_by_brand", condition = "brand = :brand")
    @Filter(name = "phones_by_phone_model", condition = "phone_model = :phone_model")
    @Filter(name = "phones_by_mobile_operator", condition = "mobile_operator = :mobile_operator")
    public List<MobilePhone> mobilePhones = new ArrayList<>();

    public static void enableFilter(String filterName, String parameterName, Object value) {
        JpaOperations.getEntityManager().unwrap(Session.class)
                .enableFilter(filterName)
                .setParameter(parameterName, value);
    }

    public static void enableFilter(String filterName, Map<String, Object> params) {
        JpaOperations.getEntityManager().unwrap(Session.class)
                .enableFilter(filterName);
        for (Map.Entry<String, Object> entry : params.entrySet()) {
            JpaOperations.getEntityManager().unwrap(Session.class)
                    .getEnabledFilter(filterName)
                    .setParameter(entry.getKey(), entry.getValue());
        }

    }

    public static PanacheQuery<Person> findByFilters(String orderBy) {
        if (orderBy != null) {
            return Person.find("select a from Person a", Sort.by(orderBy));
        }
        return Person.find("select a from Person a");
    }

    public Person add(MobilePhone mobilePhone) {
        this.mobilePhones.add(mobilePhone);
        return this;
    }
}
```

```
import com.fasterxml.jackson.annotation.JsonBackReference;
import io.quarkus.hibernate.orm.panache.PanacheEntity;
import io.quarkus.hibernate.orm.panache.PanacheQuery;
import io.quarkus.hibernate.orm.panache.runtime.JpaOperations;
import io.quarkus.panache.common.Sort;
import org.hibernate.Session;
import org.hibernate.annotations.Filter;
import org.hibernate.annotations.FilterDef;
import org.hibernate.annotations.ParamDef;

import javax.persistence.*;
import java.time.LocalDate;
import java.util.Map;

@Entity
@Table(name = "mobilephones")

@FilterDef(name = "mobile_operator", parameters = @ParamDef(name = "mobile_operator", type = "string"))
@Filter(name = "mobile_operator", condition = "mobile_operator = :mobile_operator")

@FilterDef(name = "brand", parameters = @ParamDef(name = "brand", type = "string"))
@Filter(name = "brand", condition = "brand = :brand")

@FilterDef(name = "phone_model", parameters = @ParamDef(name = "phone_model", type = "string"))
@Filter(name = "phone_model", condition = "phone_model = :phone_model")

@FilterDef(name = "person_id", parameters = @ParamDef(name = "person_id", type = "integer"))
@Filter(name = "person_id", condition = "person_id = :person_id")

@FilterDef(name = "owner_by_tax_code", parameters = @ParamDef(name = "tax_code", type = "string"))
@Filter(name = "owner_by_tax_code", condition = "person_id in (select p.id from persons p where p.tax_code = :tax_code)")

public class MobilePhone extends PanacheEntity {

    public LocalDate purchase_date;
    public String mobile_operator;
    public String brand;
    @OrderBy
    public String phone_model;
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "person_id")
    @JsonBackReference
    public Person owner;

    public static void enableFilter(String filterName, String parameterName, Object value) {
        JpaOperations.getEntityManager().unwrap(Session.class)
                .enableFilter(filterName)
                .setParameter(parameterName, value);
    }

    public static void enableFilter(String filterName, Map<String, Object> params) {
        JpaOperations.getEntityManager().unwrap(Session.class)
                .enableFilter(filterName);
        for (Map.Entry<String, Object> entry : params.entrySet()) {
            JpaOperations.getEntityManager().unwrap(Session.class)
                    .getEnabledFilter(filterName)
                    .setParameter(entry.getKey(), entry.getValue());
        }

    }

    public static PanacheQuery<MobilePhone> findByFilters(String orderBy) {
        if (orderBy != null) {
            return MobilePhone.find("select a from MobilePhone a", Sort.by(orderBy));
        }
        return MobilePhone.find("select a from MobilePhone a");
    }

    public MobilePhone owner(Person owner) {
        this.owner = owner;
        return this;
    }

    public MobilePhone purchase_date(LocalDate purchase_date) {
        this.purchase_date = purchase_date;
        return this;
    }

    public MobilePhone mobile_operator(String mobile_operator) {
        this.mobile_operator = mobile_operator;
        return this;
    }

    public MobilePhone brand(String brand) {
        this.brand = brand;
        return this;
    }

    public MobilePhone phone_model(String phone_model) {
        this.phone_model = phone_model;
        return this;
    }

    @Override
    public String toString() {
        return "MobilePhone{" +
                "purchase_date=" + purchase_date +
                ", mobile_operator='" + mobile_operator + '\'' +
                ", brand='" + brand + '\'' +
                ", phone_model='" + phone_model + '\'' +
                ", owner=" + owner +
                '}';
    }
}
```

```
import io.quarkus.hibernate.orm.panache.PanacheQuery;
import io.quarkus.panache.common.Page;
import ovh.flw.quarkusfilters.model.secondarytable.Person;

import javax.transaction.Transactional;
import javax.ws.rs.*;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.UriInfo;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;


@Path("/api/persons")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class PersonsRs {

    @Context
    UriInfo ui;

    private List<Gift> paginate(PanacheQuery<Gift> search, Integer startRow, Integer pageSize) {
        long listSize = search.count();
        if (listSize == 0) {
            return new ArrayList<>();
        } else {
            int currentPage = 0;
            if (pageSize != 0) {
                currentPage = startRow / pageSize;
            } else {
                pageSize = Long.valueOf(listSize).intValue();
            }
            return search.page(Page.of(currentPage, pageSize)).list();
        }
    }
   
    @GET
    @Transactional
    public Response getList(
            @DefaultValue("0") @QueryParam("startRow") Integer startRow,
            @DefaultValue("10") @QueryParam("pageSize") Integer pageSize,
            @QueryParam("orderBy") String orderBy) {

            if ((ui.getQueryParameters().containsKey("lastName")
                && ui.getQueryParameters().getFirst("lastName") != null)) {
                       Person.enableFilter("lastName", "lastName", "%" + get("lastName") + "%");
             }
            PanacheQuery<Person> search = Person.findByFilters(orderBy);
            List<Person> list = paginate(search, startRow, pageSize);
            return Response.ok(list).build();
        } catch (Exception e) {
            e.printStackTrace();
            return Response.serverError().build();
        }
}
```

**Comment by marceloverdijk** [comment]
Thanks for reporting this.

For me it's still not clear how this `getEntityManager().unwrap(Session.class).enableFilter(filterName)` works (as I have no real experience with it).

Is it used on only the next query or all subsequent queries? Is it shared for the whole application?
What is a exception occurs between enabling a filter and executing a query?
Probably there has been thought about these cases, but I cannot find it the docs to be honest.

I also wonder why there is a `findByFilters` method needed, as it does not seem to do anything special?

```
public static PanacheQuery<MobilePhone> findByFilters(String orderBy) {
        if (orderBy != null) {
            return MobilePhone.find("select a from MobilePhone a", Sort.by(orderBy));
        }
        return MobilePhone.find("select a from MobilePhone a");
    }
```

I could also imagine a method signature like:

`PanacheQuery<Person> livingPersons = Person.find(filter);`
`PanacheQuery<Person> livingPersons = Person.find(filter, Sort.by("name"));`

Where `filter` is a `Filter` class provided by Panache.
It would contain something like:

```
public class Filter {

    private List<Filter> filters;
}

public class FilterDetail {

    private String name; // HIbernate filter name
    private Map<String, List<Values>> parameters // Parameter with values
}
```

It could have some fluent/builder style methods just like `Sort.by` to do something like:

```
Filter f = Filter.by("myFilter", "myFilterParam", "some-value");
```

But could also support more advanced cases like:

```
Filter f = Filter.by("myFilter1")
                     .withParameter("myFilterParam1", "some-value")
                     .withParameter("myCollectionFilterParam2(List.of("foo", "bar")
                 .and("myFilter2", "anotherFilterParam", "another-value);
```

And the call:

```
Sort sort = Sort.by("name").and("birth")
PanacheQuery<Person> livingPersons = Person.find(filter, Ssort);
```

(as you could see it is also consistent with the `Sort.by` fluent api)

**Comment by FroMage** [comment]
I looked at the ORM filter docs, and it appears to me that this is really per-session, so not tied to any query, so we don't want to tie it to regular operations.

I would have said we could add the following methods in `PanacheEntityBase`:

```java
public static void enableFilter(String filterName, Parameters params);
public static void enableFilter(String filterName, Map<String,Object> params);
```

Because most examples put filter definitions on entity types, but the `@FilterDef` annotation is also allowed on packages, in which case it's probably better to put those methods in the `Panache` utility class.

At least, this API makes it consistent on how to set parameters between filters and queries, by using the same pattern of `Map/Parameters` (I've not included positional params, as those are not supported by filters).

WDY @Sanne @emmanuelbernard ?

**Comment by marceloverdijk** [comment]
But what does this mean exactly: `and it appears to me that this is really per-session, so not tied to any query` ?

**Comment by FroMage** [comment]
Yup. This is a `Session` API in ORM.

**Comment by FroMage** [comment]
It will apply to the entire transaction/request, that's what it means. You can't enable/disable it for a single query.

**Comment by marceloverdijk** [comment]
Yes I was afraid so... personally I need something per query.

**Comment by Sanne** [comment]
Yes the existing ORM API is tied to the Session, but what I'm suggesting is to have something that gives a similar experience in Panache - tied to a query.

In terms of implementation, the `Filter` API is applied on the Session just to make it easy to apply them on any Session method - not just queries. Panache could very easily enable the filter, run the query, disable the filter.

@FroMage WDYT?

**Comment by FroMage** [comment]
I can do that, but I strongly suspect that this won't work with lazy-loading, right? I assume the filter being on the session is also meant for stuff that later gets lazy-loaded.

**Comment by Sanne** [comment]
> I can do that, but I strongly suspect that this won't work with lazy-loading, right? I assume the filter being on the session is also meant for stuff that later gets lazy-loaded.

What is your concern? Since we don't want filters to impact anything beyond the query itself, it seems ok to disable them right away - so they won't filter lazy loaded objects since by definition they are loaded in a deferred way.

**Comment by marceloverdijk** [comment]
@Sanne 

>  Panache could very easily enable the filter, run the query, disable the filter.

I don't know the internals but what about possible concurrency issues?

**Comment by fiorenzino** [comment]
If i use a @Singleton Resource (SessionVerificationRs), and @inject EntityManager, 
the enabling of filter is for rest request (In the same request, you can disable the filter - to reset him - you can call:
 ```
public static void disableFilter(String filterName) {
        JpaOperations.getEntityManager().unwrap(Session.class)
                .disableFilter(filterName);
    }
```
But at next rest request, the filtering is not present.
Is not sufficient for you @marceloverdijk? 

My example code:

```
@Path("/api/session")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
@Singleton
public class SessionVerificationRs {

    @Inject
    EntityManager entityManager;


    @GET
    @Path("/multiquery")
    public Response multi(@Context UriInfo uriInfo) {
        System.out.println("----------------multiquery START----------------");
        if (uriInfo.getQueryParameters().getFirst("lastName") != null) {
            System.out.println("ENABLING lastName");
            Person.enableFilter("lastName", "lastName",
                    "%" + uriInfo.getQueryParameters().getFirst("lastName") + "%");
        }
        List<Person> list1 = Person.list("select p from Person p");
        System.out.println(list1);
        System.out.println("second query");

        Person.disableFilter("lastName");

        List<Person> list2 = Person.list("select p from Person p");
        System.out.println(list2);
        System.out.println("----------------multiquery END----------------");
        return Response.ok().build();
    }
}
```
@FroMage @Sanne if you decide to add some method to enable filter, can you evaluate also a "disableFilter" method in PanacheEntity?

**Comment by Sanne** [comment]
> @Sanne
> 
> > Panache could very easily enable the filter, run the query, disable the filter.
> 
> I don't know the internals but what about possible concurrency issues?

It's not a legal use case to use a single Session concurrently; we'll most likely detect it and throw an exception.

**Comment by FroMage** [comment]
> Panache could very easily enable the filter, run the query, disable the filter

Well hold on, then it's not even a case of wanting to put it as a method on the entity vs global. You seem to imply it should be for a single query.

But given that it's enabled on the session, it sounds dangerous because query operations can be done outside the scope of the operation that retreived them, such as not just lazy loading but also `stream()` or (very soon) reactive queries.

If we assume we want it per-query, we'd have an API a bit like:

```java
// with filter
Stream<MyEntity> ret1 = MyEntity.find("name", "stef").withFilter("myFilter", Parameters.with("param", true)).stream();
// no filter
Stream<MyEntity> ret2 = MyEntity.find("name", "stef").stream();
ret.forEach();// we've already removed the filter from the session at this point
```

Perhaps there are ways to turn this back on by writing a `Stream` wrapper, but this is starting to look like fitting circles into squares.
We can also decide to make it work only in certain scenarios, and tell people not to use streams or lazy loading when using filters.

Ultimately, I guess it becomes a matter of usability:

- We can justify limitations in how this works if you tell me filters should be per-query (and the per-session API in Hibernate was a bad choice in retrospect), and make it work for some use-cases (hopefully the most common ones)
- Or if you tell me it really belongs per-session, I can add either `MyEntity.enableFilter(name, params)` (fake that it's per entity) or `Panache.enbleQuery(name, params)` (admit that it's global).

**Comment by Sanne** [comment]
> But given that it's enabled on the session, it sounds dangerous because query operations can be done outside the scope of the operation that retreived them, such as not just lazy loading but also stream() or (very soon) reactive queries.

That is a good point, but I don't think it will be a problem: what matters is if the filter was enabled (or not) when the select criteria is being rendered. Even if you're streaming results, it's sent only once to the DB when the resultset is being created.

> We can justify limitations in how this works if you tell me filters should be per-query (and the per-session API in Hibernate was a bad choice in retrospect), and make it work for some use-cases (hopefully the most common ones)

I think the API is fine as it is, as it can be used for many other cool uses *as well*.
For example:
 - you can restrict the data the logged-in-user can see at a session level by enabling some filters from a security context.
 - you can implement various forms of multi-tenancy, in particular those in which all tenants are storing data in the same DB and schema , while you restrict what each tenant can do via some `tenantId` column in a selection of tables. Most useful when you only want to separate by tenants a subset of your model while retaining referential integrity and 2LC across the shared model.

Yet the fact that this feature is quite powerful doesn't disualify it from making single queries more maintainable when there's many dynamic restrictions.

> We can justify limitations in how this works

Well which limitations :)  I'm sure there might be some bugs I don't know about as apparently not many know about this capability, but I'm not aware of any and it served me very well in the past.

> Or if you tell me it really belongs per-session, I can add either MyEntity.enableFilter(name, params) (fake that it's per entity) or Panache.enbleQuery(name, params) (admit that it's global).

Yes that's a possibility, but it depends what you mean to address. It would certainly be nice to have the `Panache.enableFilter` / `disableFilter` capabilities, but also I believe we should have something which is explicitly more friendly for dynamic filtering of a query. You have a better alternative in mind?

**Comment by FroMage** [comment]
> That is a good point, but I don't think it will be a problem: what matters is if the filter was enabled (or not) when the select criteria is being rendered. Even if you're streaming results, it's sent only once to the DB when the resultset is being created.

OK, in that case I believe it's fine. I thought it would be much more dangerous.

> I think the API is fine as it is, as it can be used for many other cool uses as well.

OK, now I understand it really serves more than single queries. So we actually need two APIs, right? One for turning this on for a single query, and another for turning it on globally.

Now questions:

- If I put the global `enableFilter` on the entity, we're going to not be able to enable filters declared on packages. I can add an extra `Panache.enableFilter` for those. Or only the latter? User-wise, how common are filters on entities vs on packages?
- If I add `Entity.find("name", "stef").withFilter("filterName" ...)` can I restore the `filterName` setting to what value it had before once the query is done? If the user had it on globally, and I disable it after the query, they might be puzzled. Can I query Hibernate to know if a given filter is enabled or not, and get its parameters (I assume enabling it back will need its previous parameters)?

**Comment by emmanuelbernard** [comment]
Can I play the grumpy person?

Does this feature look like a Panache 80% use case simplification to you. Or more a 20% advanced use case?
Are we going to make Panache as complete as JPA? This is not the original intent.

Are we losing much by requesting people to do

```
Person.getEntityManager().unwrap(Session.class).enableFilter("foo");
```

**Comment by Sanne** [comment]
> So we actually need two APIs, right? One for turning this on for a single query, and another for turning it on globally.

Yes I believe differentiating the API would make it clearer to users how to accomplish such things. That probably wasn't good enough in the ORM upstream API.

However, I think you have the option to expose only one of these. What makes most sense to expose in Panache? That's up to you.

> User-wise, how common are filters on entities vs on packages?

I have no idea. More importantly, I believe it doesn't matter: we probably don't want this to be entity-scoped as I don't think you can "emulate" this in the most correct sense. Let's stick on making dynamic restrictions on queries better expressed?

> can I restore the filterName setting to what value it had before once the query is done? If the user had it on globally, and I disable it after the query, they might be puzzled. Can I query Hibernate to know if a given filter is enabled or not

Yes it's possible to know which filters are enabled. Not sure if we can extract parameters; worst case we can change ORM to expose this but I'd argue we might prefer to throw an exception if the same filter is being applied multiple times? I doubt it would be correct to apply the same filter multiple times.

**Comment by Sanne** [comment]
I think @emmanuelbernard has a point. Indeed I wasn't advocating to expose those filters globally - those additional use cases I mentioned are just an explanation about why the API is what it is on ORM, since @FroMage asked.   Please don't see my explanation as a request to expose all such nuances on Panache.

I still think though that a filter approach is how it should look like to "stack" additional restrictions on the query, especially when there's customization at play which are exposed to users: it needs to be easy to add restrictions dymanically, yet it needs to be highly defensive so to not encourage appending elements from the user UI. (security issues et al).

Also, it would be nice to have a "core query" which can be actually rather complex, such as needing to be expressed in HQL - and yet be able to be applied filters on top dynamically without needing to mess with the HQL source (and force constant reparsing at runtime).

**Comment by FroMage** [comment]
OK, so the advice would be to add the filter methods on `PanacheQuery` initially and only there and see where this leads us.

**Comment by FroMage** [comment]
PR at https://github.com/quarkusio/quarkus/pull/9238
